Update cnames after getting certs

tlwr opened the pull request:
What
----

Delay setting CNAMEs until IAM has a valid certificate for the distribution

Why
---

Amazon [have recently changed](https://aws.amazon.com/about-aws/whats-new/2019/04/amazon-cloudfront-enhances-the-security-for-adding-alternate-domain-names-to-a-distribution/) how CloudFront works. Now a valid certificate for any and all alternate names of the distribution are required.

How
---

@bengerman13 has kindly done some of the hard work for us upstream, this PR cherry-picks the commit in. 🙇 👏 

As Miki reports the Update step requires an extra fiddle which is in another commit.

How to review
---

- Code review
- I'm going to run this through my development environment and do some manual testing

bengerman13 committed:
Update cnames after getting certs

CloudFront recently changed their requirements so we now need to upload
a certificate whenever setting a distribution cname. This PR delays
setting the cname until we have a certificate.

46bit commented:
Provisioning now works. But updating to add an extra domain still fails:

```
cf update-service rmi-paas-cdn-bobtest -c '{"domain": "paas-test.toby.codes,cdn.46bit.com"}'
Updating service instance rmi-paas-cdn-bobtest as admin...
FAILED
Server error, status code: 502, error code: 10001, message: Service broker error: InvalidViewerCertificate: The certificate that is attached to your distribution doesn't cover the alternate domain name (CNAME) that you're trying to add. For more details, see: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/CNAMEs.html#alternate-domain-names-requirements
	status code: 400, request id: cd006f5c-603d-11e9-9969-cb860e235efc
```

Let's fix updating and we can gift that to 18F in return for [their provisioning fix](https://github.com/18F/cf-cdn-service-broker/pull/116) 😄

Commit:
Use old domains when updating CloudFront

Signed-off-by: Toby Lorne <toby.lornewelch-richards@digital.cabinet-office.gov.uk>

tlwr commented:
Create a service
```
 ~/Code/paas-cf  ⑂ wip-tlwr  Tue 16 Apr 2019 11:05:27 BST (tlwr) ✅
🧢   cf create-service cdn-route cdn-route rmi-paas-cdn-bobtest -c '{"domain": "paas-test.toby.codes"}'
Creating service instance rmi-paas-cdn-bobtest in org govuk-paas / space tools as admin...
OK

Create in progress. Use 'cf services' or 'cf service rmi-paas-cdn-bobtest' to check operation status.
 ~/Code/paas-cf  ⑂ wip-tlwr  Tue 16 Apr 2019 11:05:43 BST (tlwr) ✅
🧢   cf service rmi-paas-cdn-bobtest
Showing info of service rmi-paas-cdn-bobtest in org govuk-paas / space tools as admin...

name:            rmi-paas-cdn-bobtest
service:         cdn-route
tags:
plan:            cdn-route
description:     Custom domains, CDN caching, and TLS certificates with automatic renewal
documentation:   https://cloud.gov/docs/services/cdn-route/
dashboard:

Showing status of last operation from service rmi-paas-cdn-bobtest...

status:    create in progress
message:   Provisioning in progress [paas-test.toby.codes => tlwr.dev.cloudpipelineapps.digital]; CNAME or ALIAS domain paas-test.toby.codes to d96ohnnfy4ebt.cloudfront.net and create TXT record(s):
name:
           _acme-challenge.paas-test.toby.codes., value: XmIWp8gcj0xeZxPuivoGaqQSexEO3OhsQP6R5lChKbE, ttl: 120
started:   2019-04-16T10:05:40Z
updated:   2019-04-16T10:05:42Z

There are no bound apps for this service.
```
Create succeeds
then
Update a service
```
 ~/go/src/github.com/18F/cf-cdn-service-broker  ⑂ apply-upstream-cdn-patch  Tue 16 Apr 2019 14:00:45 BST (tlwr) ✅
🧢   cf update-service rmi-paas-cdn-bobtest -c '{"domain": "paas-test.toby.codes,paas-test-2.toby.codes"}'
Updating service instance rmi-paas-cdn-bobtest as admin...
OK

Update in progress. Use 'cf services' or 'cf service rmi-paas-cdn-bobtest' to check operation status.
```
Update succeeds
```
 ~/go/src/github.com/18F/cf-cdn-service-broker  ⑂ cname-after-certs ⇡1  Tue 16 Apr 2019 14:09:19 BST (tlwr) ✅
🧢   cf service rmi-paas-cdn-bobtest
Showing info of service rmi-paas-cdn-bobtest in org govuk-paas / space tools as admin...

name:            rmi-paas-cdn-bobtest
service:         cdn-route
tags:
plan:            cdn-route
description:     Custom domains, CDN caching, and TLS certificates with automatic renewal
documentation:   https://cloud.gov/docs/services/cdn-route/
dashboard:

Showing status of last operation from service rmi-paas-cdn-bobtest...

status:    update succeeded
message:   Service instance provisioned [paas-test.toby.codes,paas-test-2.toby.codes => tlwr.dev.cloudpipelineapps.digital]; CDN domain
           d96ohnnfy4ebt.cloudfront.net
started:   2019-04-16T13:01:02Z
updated:   2019-04-16T13:10:49Z

There are no bound apps for this service.
```

tlwr commented:
👀 ➡️ https://paas-test.toby.codes/
👀 ➡️ https://paas-test-2.toby.codes/

richardTowers commented on utils/cloudfront.go:
```diff
@@ -259,6 +260,33 @@ func (d *Distribution) Get(distId string) (*cloudfront.Distribution, error) {
+	DistributionConfig.ViewerCertificate.IAMCertificateId = aws.String(certId)
+	DistributionConfig.ViewerCertificate.CertificateSource = aws.String("iam")
+	DistributionConfig.ViewerCertificate.SSLSupportMethod = aws.String("sni-only")
+	DistributionConfig.ViewerCertificate.MinimumProtocolVersion = aws.String("TLSv1.2_2018")
```
Should this be hardcoded? Is this the same `MinimumProtocolVersion` we have at the moment?

46bit replied:
It does match what we set in https://github.com/alphagov/paas-cdn-broker/pull/12/files#diff-9ce0b3b0fd6c33664bdb1b8f843fa74eR276. I'll see about having that in one place.

46bit replied:
Ah, this function entirely replaces the `SetCertificate` where the other occurence is. Will remove `SetCertificate`.

richardTowers reviewed (approved):
Code looks good to me, and it can't get any more broken than it currently is. One comment about a hardcoded string though...

46bit commented:
I've tried adding new domains, removing domains, etc successfully. This change seems sound.

Commit:
Remove unused Cloudfront utils code

`SetCertificateAndCname` replaced all uses of `SetCertificate` in
commit 66c7f8912c18970acaf4d816c4d5dc9883192649. We are unlikely to
use this code again because of the changes in AWS that caused that
commit to be made.
